### DISCUSSION TITLE
How to pass the "current frame" when duckdb is wrapped (by a Jupyter magic or other function)

### DISCUSSION THREAD

**Discussion by paultiq** [duckdb/duckdb#14119] [General]
The recent change in https://www.github.com/duckdb/duckdb/pull/12425 changed the default behavior to only inspect the current frame's locals(). This can be overridden to older behavior in https://www.github.com/duckdb/duckdb/pull/13896. 

This is a great change, including for the reasons outlined in the original issue: https://github.com/duckdb/duckdb/issues/11687.

However, it impacts any code that wraps duckdb. [magic_duckdb](<https://github.com/iqmo-org/magic_duckdb>) is a Jupyter magic (cell & line) that wraps duckdb:
```py
%load_ext magic_duckdb
mydf = pd.DataFrame(...)
%dql select * from mydf
``` 

The above code works fine in <=1.0.0, but in >=1.1.0 won't because the %dql magic's frame is two frames "lower" than the callers frame. 

My question / challenge is: how to preserve the "current frame" behavior (the new default) from cell magics, or any other function that wraps duckdb? 

### Workaround 1:
`python_scan_all_frames`: This works, but is suboptimal due to reasons mentioned in the original change. It makes sense to only scan the current frame. 

### Workaround 2: Patch f_locals
One workaround that works is to reach back and grab the "right" frame: 
`context = inspect.currentframe().f_back.f_back.f_locals # type: ignore`
and then push those values into the current f_locals: 
`inspect.currentframe().f_locals.update(context)`

This is not a great solution, for a variety of reasons, also because the behavior of modify f_locals in this way is not documented. 

### Ideas
A few ideas: 
1. Allow passing of a dictionary of name/object pairs to use, instead of f_locals. This gives the caller explicitly control... this would basically be like implicit registration: instead of having to register all the DF's you use, you could pass a dictionary of them, and they'd be used if needed.
2. Add a parameter for stack depth (whether last N or an explicit frame reference: get locals from 3 frames up)
3. Or, Extract the inspect logic to python code (out of the C code) so it can be extended or monkey patched. Because it's implemented in C, I couldn't just override the behavior.

**Comment by Tishj** [comment]
I don't have a satisfying answer for this, other than that I've had some success with `eval`, that allows you to provide locals and globals to execute with, which can be retrieved with `currentframe().f_back ..` like you mentioned

**Reply by paultiq** [reply]
What do you think about allowing a dictionary to be passed to use -in place- of f_locals? This has added benefit of implicit registration: passing a dictionary of objects without having to register them all.

Alternatively: 
Or, moving the lookup function into the Python layer so it can be monkey patched? (Not sure this is feasible)

Having to modify f_locals is not good, but I can't patch because of where the logic is implemented... even tho it's just a simple inspect/etc.

**Comment by Tishj** [comment]
I haven't heard that third suggestion before, I don't hate it
maybe we can make a python method that returns a list of found objects, which are then tried in-order by the replacement scan logic.

Monkey patching wouldn't work, but we could create a method to set an override, which is used by the system when it's set.

**Reply by paultiq** [reply]
Maybe accept a callable, such that the default is `lambda x: inspect.currentframe().f_locals` but another callable could be used per request? Which would allow a fixed context or a different frame logic?

**Comment by Tishj** [comment]
the first suggestion is pretty much what <https://github.com/duckdb/duckdb/discussions/10191> suggests

the second suggestion could also work as a more limited solution, `scan_all_frames` could be reworked to be a number of frames instead
